UN-3040 Get token counting working for models other than OpenAI and Anthropic

Noravee opened the pull request.

Noravee commented on neuro_san/internals/run_context/langchain/get_ollama_callback.py:
```diff
@@ -0,0 +1,39 @@
+
+@contextmanager
+def get_ollama_callback() -> Generator[OllamaCallbackHandler, None, None]:
+    """Get ollama callback.
```
Context manager for Ollama and any other models.

d1donlydfink replied:
Maybe add a comment with a pointer to langchain code that you based this off of.

Noravee replied:
Comment added

Noravee commented on neuro_san/internals/run_context/langchain/ollama_callback_handler.py:
```diff
@@ -0,0 +1,122 @@
+
+class OllamaCallbackHandler(BaseCallbackHandler):
+    """
+    Callback handler that tracks token usage via "AIMessage.usage_metadata".
```
Callback handler for Ollama and any other models. Work for any chat model from langchain-core >=0.2.2.

Noravee replied:
Maybe "OllamaCallbackHandler" is misleading because this works with any llm. Change this to "LlmCallbackHandler".

d1donlydfink replied:
Not a bad idea.

Noravee commented on neuro_san/internals/run_context/langchain/langchain_token_counter.py:
```diff
@@ -253,7 +259,6 @@ def normalize_token_count(callback: BaseCallbackHandler, time_taken_in_seconds:
                 "total_cost": callback.total_cost,
                 "time_taken_in_seconds": time_taken_in_seconds,
                 "caveats": [
-                    "Only doing token accounting for OpenAI, AzureOpenAI and Anthropic models for now.",
```
Remove "Only doing token accounting for OpenAI, AzureOpenAI and Anthropic models for now.".

Noravee commented on neuro_san/internals/run_context/langchain/langchain_token_counter.py:
```diff
@@ -209,7 +212,7 @@ def get_callback_for_llm(llm: BaseLanguageModel) -> Any:
             return get_bedrock_anthropic_callback
 
-        return None
+        return get_ollama_callback
```
Use this callback for any models other than OpenAI and Anthropic.

Noravee commented on neuro_san/internals/run_context/langchain/ollama_callback_handler.py:
```diff
@@ -0,0 +1,122 @@
+    completion_tokens: int = 0
+    successful_requests: int = 0
+    # There is no cost for Ollama model. Define "total_cost" and set at 0.0 to fit the report template.
+    total_cost: float = 0.0
```
"total_cost" is fixed at 0.0

d1donlydfink replied:
Maybe just not include total_cost at all?
Or report None if it's necessary?
Maybe there's some super-class that requires some kind of number?

Noravee replied:
There is no requirement from superclass, but I think we should keep it since there could be a mixture of models in an agent network or llm fallback. We could start with Ollama which has no cost but then fallback to gpt-4o.

The current reporting structure does not contain model name so we don't know the token usage of each model. I think later we can change the report structure to
```
{
            "gpt-4o-mini-2024-07-18": {
                "input_tokens": 8,
                "output_tokens": 10,
                "total_tokens": 18,
                "successful_requests": 2,
                "total_cost":  0.0001
            },
            "claude-3-5-haiku-20241022": {
                "input_tokens": 8,
                "output_tokens": 21,
                "total_tokens": 29,
                "successful_requests": 2,
                "total_cost":  0.0001
            },
        }
```

d1donlydfink commented on neuro_san/internals/run_context/langchain/langchain_token_counter.py:
```diff
@@ -226,7 +229,7 @@ def get_context_var_for_llm(llm: BaseLanguageModel) -> ContextVar:
             return bedrock_anthropic_callback_var
 
-        return None
+        return ollama_callback_var
```
Should we be checking for an instance type of ChatOllama or is this really going to take care of all other cases?
If all other cases, then add a comment to that effect here and maybe above too.

Noravee replied:
Maybe "OllamaCallbackHandler" is misleading because this works with any llm. Change this to "LlmCallbackHandler".

d1donlydfink commented on neuro_san/internals/run_context/langchain/ollama_callback_handler.py:
```diff
@@ -0,0 +1,122 @@
+
+    Note: Ollama models currently have no associated cost, so "total_cost" is always set to 0.0 for compatibility with
+    reporting templates.
+    """
```
This is a very nice comment!

d1donlydfink commented on neuro_san/internals/run_context/langchain/ollama_callback_handler.py:
```diff
@@ -0,0 +1,122 @@
+    def __init__(self) -> None:
+        """Initialize the CallbackHandler."""
+        super().__init__()
+        self._lock = threading.Lock()
```
This is run in asynchronous code. You should probably use asyncio.Lock() instead

Noravee replied:
I have changed it to `asyncio.Lock()` and change the superclass to `AsyncCallbackHandler`.

FWIW, both `OpenAICallbackHandler` and `BedrockAnthropicTokenUsageCallbackHandler` use `threading.Lock()` and `BaseCallbackHandler` which are not async code.

d1donlydfink reviewed (commented):
Please look at in-code comments.

Also: How did you test with and without the feature (to be sure you didn't break anything)?

Noravee commented:
Thank you for the feedback. I've made the following updates:

Replaced "Ollama" with "Llm_Token" since it can be used across different LLMs.

Added a reference to the context_manager module.

Updated ``Llm_Token_Callback_Handler`` to use async code:

Switched from ``BaseCallbackHandler`` to ``AsyncCallbackHandler``

Replaced ``threading`` with ``asyncio``

**Testing**:
I tested by running ``website_search_usage_example.hocon``, which internally calls ``website_search.hocon`` as a subnetwork. I verified the feature with the following model combinations to ensure nothing broke:

``gpt-4o``, ``gpt-4o``

``gpt-4o``, ``ollama3.1``

``ollama3.1``, ``ollama3.1``

Please let me know if anything else needs clarification or adjustment.
